### PR TITLE
Switch to 30-bit digits

### LINKED ISSUE #40: Consider switching internal storage to 30 bits per digit
As of Chrome 80 (V8 8.0 branch), V8's "pointer compression" technique has required it to use "31-bit Smis" under the hood, which means that JSBI's chosen internal representation of 32 bits per digit is now a lot less efficient than it was before (in terms of both memory and CPU). It might be worth switching to 30 bits per digit, if enough people care about JSBI's efficiency (when it isn't compiled to native BigInts with the [Babel plugin](https://github.com/GoogleChromeLabs/babel-plugin-transform-jsbi-to-bigint)).

(As an alternative, we could move to 50 or so bits per digit to at least make more efficient use of doubles, but my intuition is that the required double<->int conversions would be too costly.)

### PR DESCRIPTION
**Pull request by jakobkummerow**
The primary motivation is that this is more efficient on recent
versions of V8, where pointer compression is enabled and "Smis"
hence only have 31 bits of payload.
A secondary benefit is that additions and subtractions can be
done on full digits rather than half-digits, because we can check
for carry/borrow bits on the 32-bit intermediate result.

This shaves around 30-40% off the time taken for most operations,
and saves some memory for JSBI.BigInt instances too.

Fixes #40.

Detailed performance results: https://docs.google.com/spreadsheets/d/1Z1EVZ9DM1S8IGKzxvRdQzCpXEDLRHo5YROEMeTYOq8I/edit?usp=sharing

### CODE REVIEW

File: jsbi.mjs [hunk 1]
```diff
@@ -393,13 +394,14 @@ class JSBI extends Array {
...
+    n = Math.floor(n);
     if (n === 0) return JSBI.__zero();
     // If {x} has less than {n} bits, return it directly.
     if (n >= JSBI.__kMaxLengthBits) return x;
-    const neededLength = (n + 31) >>> 5;
+    const neededLength = ((n + 29) / 30) | 0;
     if (x.length < neededLength) return x;
     const topDigit = x.__unsignedDigit(neededLength - 1);
-    const compareDigit = 1 << ((n - 1) & 31);
+    const compareDigit = 1 << ((n - 1) % 30);
```
Comment (mathiasbynens) [inline jsbi.mjs:404]: Same here

File: jsbi.mjs [hunk 2]
```diff
@@ -429,9 +432,9 @@ class JSBI extends Array {
     }
     // If {x} is positive and has up to {n} bits, return it directly.
     if (n >= JSBI.__kMaxLengthBits) return x;
-    const neededLength = (n + 31) >>> 5;
+    const neededLength = ((n + 29) / 30) | 0;
     if (x.length < neededLength) return x;
-    const bitsInTopDigit = n & 31;
+    const bitsInTopDigit = n % 30;
```
Comment (mathiasbynens) [inline jsbi.mjs:437]: …and here

File: jsbi.mjs [hunk 3]
```diff
@@ -584,14 +587,14 @@ class JSBI extends Array {
...
-    const digits = (exponent >>> 5) + 1;
+    const digits = ((exponent / 30) | 0) + 1;
     const result = new JSBI(digits, sign);
     const kHiddenBit = 0x00100000;
     let mantissaHigh = (JSBI.__kBitConversionInts[1] & 0xFFFFF) | kHiddenBit;
     let mantissaLow = JSBI.__kBitConversionInts[0];
     const kMantissaHighTopBit = 20;
     // 0-indexed position of most significant bit in most significant digit.
-    const msdTopBit = exponent & 31;
+    const msdTopBit = exponent % 30;
```
Comment (mathiasbynens) [inline jsbi.mjs:597]: and here

**Review by mathiasbynens — APPROVED** [review]
Rubberstamp'd with some questions

File: jsbi.mjs [hunk 4]
```diff
@@ -177,12 +178,12 @@ class JSBI extends Array {
...
     if (x.length === 1 && x.__digit(0) === 2) {
       // Fast path for 2^n.
-      const neededDigits = 1 + (expValue >>> 5);
+      const neededDigits = 1 + ((expValue / 30) | 0);
       const sign = x.sign && ((expValue & 1) !== 0);
       const result = new JSBI(neededDigits, sign);
       result.__initializeDigits();
       // All bits are zero. Now set the n-th bit.
-      const msd = 1 << (expValue & 31);
+      const msd = 1 << (expValue % 30);
```
[the remark this answers is not in the document]
Comment (jakobkummerow) [reply jsbi.mjs:186]: Yes, intentional. `&31` is the optimized version of saying `%32`. As we're switching from 32-bit digits to 30-bit digits, all "divide number of bits by bits-per-digit to get number of digits" operations are switching from `%32` to `%30`. Unfortunately there is no simple bitwise operation to express a division by 30, the `&` trick only works for powers of two.

File: jsbi.mjs (hunk 1 above)
Comment (jakobkummerow) [reply jsbi.mjs:404]: Same answer here and twice more below.

**Comment by jakobkummerow** [comment]
Thanks for the review! Merging.

**Pull request merged**